languages/python/sqlalchemy-oso/tests/test_post_relationship.py

```diff
@@ -0,0 +1,530 @@
+"""Standardized tests for adapters based on the Post model.
+
+Tests come from the relationship document & operations laid out there
+https://www.notion.so/osohq/Relationships-621b884edbc6423f93d29e6066e58d16.
+"""
+import pytest
+
+from sqlalchemy import create_engine
+from sqlalchemy.engine import Engine
+from sqlalchemy.orm import sessionmaker, relationship
+from sqlalchemy.orm.session import Session
+from sqlalchemy.schema import Table
+from sqlalchemy.ext.declarative import declarative_base
+
+from sqlalchemy import Column, Integer, String, Enum, Boolean, ForeignKey
+
+from oso import Oso
+from sqlalchemy_oso.auth import authorize_model, register_models
+
+ModelBase = declarative_base(name="ModelBase")
+
+
+def print_query(query):
+    print(query.statement.compile(), query.statement.compile().params)
+
+
+class Tag(ModelBase):
+    __tablename__ = "tags"
+
+    name = Column(String, primary_key=True)
+    created_by_id = Column(Integer, ForeignKey("users.id"))
+    created_by = relationship("User", foreign_keys=[created_by_id])
+
+    users = relationship("User", secondary="user_tags")
+
+    # If provided, posts in this tag always have the public access level.
+    is_public = Column(Boolean, default=False, nullable=False)
+
+
+post_tags = Table(
+    "post_tags",
+    ModelBase.metadata,
+    Column("post_id", Integer, ForeignKey("posts.id")),
+    Column("tag_id", Integer, ForeignKey("tags.name")),
+)
+
+user_tags = Table(
+    "user_tags",
+    ModelBase.metadata,
+    Column("user_id", Integer, ForeignKey("users.id")),
+    Column("tag_id", Integer, ForeignKey("tags.name")),
+)
+
+
+class Post(ModelBase):
+    __tablename__ = "posts"
+
+    id = Column(Integer, primary_key=True)
+    contents = Column(String)
+    access_level = Column(Enum("public", "private"), nullable=False)
```

Comment:
There's some slight differences here to keep in mind while aligning the test suites, like `access_level` defaulting to `"private"` in the Django test model.

Reply:
Thanks good point